src/probe_modules/module_dns.c

```diff
@@ -0,0 +1,1043 @@
+    // -- a ptr, which terminates
+    // -- a null byte, which terminates
+    // -- a segment length which either terminates or ensures we keep looping
+
+    assert(0);
+    return 0;
+}
+
+// data: Where we are in the dns payload
+// payload: the entire udp payload
+char* _get_name(char* data, uint16_t data_len, char* payload, 
+        uint16_t payload_len, uint16_t* bytes_consumed)
+{
+    log_trace("dns", "call to _get_name, data_len: %d", data_len);
+
+    char* name = xmalloc(MAX_NAME_LENGTH);
+    memset(name, 0x00, MAX_NAME_LENGTH);
+    
+    *bytes_consumed = _get_name_helper(data, data_len, payload, 
+            payload_len, name, MAX_NAME_LENGTH - 1, 0);
+
+    if (*bytes_consumed == 0) {
+        free(name);
+        return NULL;
+    }
+
+    // Our memset ensured null byte.
+    assert(name[MAX_NAME_LENGTH - 1] == '\0');             
+
+    log_trace("dns", "return success from _get_name, bytes_consumed: %d, string: %s", *bytes_consumed, name);
+
+    return name;
+
+}
+
+static bool _process_response_question(char **data, uint16_t* data_len, char* payload,
+        uint16_t payload_len, fieldset_t* list)
+{   
+    // Payload is the start of the DNS packet, including header
+    // data is handle to the start of this RR
+    // data_len is a pointer to the how much total data we have to work with.
+    // This is awful. I'm bad and should feel bad.
+    uint16_t bytes_consumed = 0;
+
+    char* question_name = _get_name(*data, *data_len, payload, payload_len,  &bytes_consumed);
+
+    // Error.
+    if (question_name == NULL) {
+        return 1;
+    }
+
+    assert(bytes_consumed > 0);
+
+    if ( (bytes_consumed + sizeof(dns_question_tail)) > *data_len) {
+        free(question_name);
+        return 1;
+    }
+
+    dns_question_tail* tail = (dns_question_tail*)(*data + bytes_consumed);
+
+    uint16_t qtype = ntohs(tail->qtype);
+    uint16_t qclass = ntohs(tail->qclass);
+    
+    // Build our new question fieldset
+    fieldset_t *qfs = fs_new_fieldset(); 
+    fs_add_string(qfs, "name", question_name, 1);
+    fs_add_uint64(qfs, "qtype", qtype);
+    if (qtype > MAX_QTYPE || qtype_qtype_to_strid[qtype] == BAD_QTYPE_VAL) {
+        fs_add_string(qfs, "qtype_str", (char*) BAD_QTYPE_STR, 0);
+    } else {
+        // I've written worse things than this 3rd arg. But I want to be fast.
+        fs_add_string(qfs, "qtype_str", (char*)qtype_strs[qtype_qtype_to_strid[qtype]], 0);
+    }
+    fs_add_uint64(qfs, "qclass", qclass);
+
+    // Now we're adding the new fs to the list.
+    fs_add_fieldset(list, "question", qfs);
+
+    // Now update the pointers.
+    *data = *data + bytes_consumed + sizeof(dns_question_tail);
+    *data_len = *data_len - bytes_consumed - sizeof(dns_question_tail);
+
+    return 0;
+}
+
+// XXX This should be merged with _process_response_question. 
+// Ended up being almost exactly the same
+bool _process_response_answer(char **data, uint16_t* data_len, char* payload,
```

Comment:
Fixed static.

Honestly. The two are different enough now that I'm fine with them being separate unless there's huge pushback.